x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptorsIntersection.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.authz;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.XContentParserUtils;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public record RoleDescriptorsIntersection(List<List<RoleDescriptor>> roleDescriptorsList) implements ToXContentObject, Writeable {
+
+    public RoleDescriptorsIntersection(StreamInput in) throws IOException {
+        this(RoleDescriptorsIntersection.readRoleDescriptorsListFrom(in));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(roleDescriptorsList.size());
+        for (List<RoleDescriptor> roleDescriptors : roleDescriptorsList) {
+            out.writeList(roleDescriptors);
+        }
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startArray();
+        {
+            for (List<RoleDescriptor> roleDescriptors : roleDescriptorsList) {
+                builder.startObject();
+                for (RoleDescriptor roleDescriptor : roleDescriptors) {
+                    builder.field(roleDescriptor.getName(), roleDescriptor);
+                }
+                builder.endObject();
+            }
+        }
+        builder.endArray();
+        return builder;
+    }
+
+    public static RoleDescriptorsIntersection fromXContent(XContentParser p) throws IOException {
+        XContentParser.Token token = p.currentToken();
+        if (token == null) {
```

Comment:
This can happen when you need parse this object on its own, i.e. when it is not part of other document. For now, this is only possible in tests. But I think having it is future proof. Also this pattern is used in many other places so I decided to have it here instead having it in test code.

Reply:
Nice, thanks for explanation. Agreed that it makes sense here.